<a href="https://colab.research.google.com/github/oliveralbrecht94/ATDIT2/blob/master/Task%20C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import needed classes

In [2]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.manifold import TSNE
from scipy import stats
import scipy.spatial.distance

np.random.seed(1337)
plt.set_cmap('gray')

<Figure size 432x288 with 0 Axes>

# Load MNIST Dataset

In [3]:
train = tf.keras.datasets.mnist.load_data()[0]
X_train, Y_train = train[0], train[1]

11493376/11490434 [==============================] - 0s 0us/step


In [0]:
n = X_train.shape[0]
m = X_train.shape[1]**2

X_train = X_train.reshape([n, m])

# Take subsamples

In [0]:
n_sample = 10000
idx = np.random.randint(n, size=n_sample)
x_sample = X_train[idx]
y_sample = Y_train[idx]



# Init Centroids

In [6]:
k = 10
# init centroids
idx_centroid = np.random.randint(n_sample, size=k)
centroids = x_sample[idx_centroid]


idx_centroid

array([3951, 5479, 6465, 7633, 1707, 3634, 6262, 6001, 7776, 7732])

In [7]:
# calculate distances
dists = scipy.spatial.distance.cdist(x_sample, centroids, metric='euclid')
dists.shape

(10000, 10)

In [8]:
#find nearest ones
labels = np.argmin(dists, axis=1)
labels

array([1, 7, 5, ..., 1, 9, 3])

In [9]:
np.unique(labels)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [0]:
# get new centroids
list = [x_sample[labels == i].mean(0) for i in range(k)]
centroids = np.stack(list)

In [0]:
# do the iteraions of kmean
#iteration = 10
#iteration = 100
iteration = 1000
for j in range(iteration):
    
    # calc distances
    dists = scipy.spatial.distance.cdist(x_sample, centroids, metric = 'euclid')
    
    # label according to nearest centroids
    centroid_labels = np.argmin(dists, axis = 1)
    
    # get new centroids
    centroid_list = [x_sample[centroid_labels == i].mean(0) for i in range(k)]
    centroids = np.stack(centroid_list)

# Clustering the data points

In [0]:
cluster_0 = x_sample[labels == 0]
cluster_1 = x_sample[labels == 1]
cluster_2 = x_sample[labels == 2]
cluster_3 = x_sample[labels == 3]
cluster_4 = x_sample[labels == 4]
cluster_5 = x_sample[labels == 5]
cluster_6 = x_sample[labels == 6]
cluster_7 = x_sample[labels == 7]
cluster_8 = x_sample[labels == 8]
cluster_9 = x_sample[labels == 9]


In [0]:
labels_cluster_0 = y_sample[centroid_labels == 0]
labels_cluster_1 = y_sample[centroid_labels == 1]
labels_cluster_2 = y_sample[centroid_labels == 2]
labels_cluster_3 = y_sample[centroid_labels == 3]
labels_cluster_4 = y_sample[centroid_labels == 4]
labels_cluster_5 = y_sample[centroid_labels == 5]
labels_cluster_6 = y_sample[centroid_labels == 6]
labels_cluster_7 = y_sample[centroid_labels == 7]
labels_cluster_8 = y_sample[centroid_labels == 8]
labels_cluster_9 = y_sample[centroid_labels == 9]

In [24]:
print("Numbers in Cluster 0: "+str(np.unique(labels_cluster_0)))
print("Numbers in Cluster 1: "+str(np.unique(labels_cluster_1)))
print("Numbers in Cluster 2: "+str(np.unique(labels_cluster_2)))
print("Numbers in Cluster 3: "+str(np.unique(labels_cluster_3)))
print("Numbers in Cluster 4: "+str(np.unique(labels_cluster_4)))
print("Numbers in Cluster 5: "+str(np.unique(labels_cluster_5)))
print("Numbers in Cluster 6: "+str(np.unique(labels_cluster_6)))
print("Numbers in Cluster 7: "+str(np.unique(labels_cluster_7)))
print("Numbers in Cluster 8: "+str(np.unique(labels_cluster_8)))
print("Numbers in Cluster 9: "+str(np.unique(labels_cluster_9)))

Numbers in Cluster 0: [0 2 3 4 5 6 8 9]
Numbers in Cluster 1: [2 3 4 5 7 8 9]
Numbers in Cluster 2: [0 1 2 3 4 5 6 7 8 9]
Numbers in Cluster 3: [0 1 2 3 4 5 6 7 8 9]
Numbers in Cluster 4: [0 2 3 5 6 7 8 9]
Numbers in Cluster 5: [0 1 2 3 5 6 8 9]
Numbers in Cluster 6: [0 1 2 3 4 5 6 7 8 9]
Numbers in Cluster 7: [0 2 3 4 5 6 8 9]
Numbers in Cluster 8: [0 1 2 3 4 5 6 7 8]
Numbers in Cluster 9: [0 1 2 3 4 5 6 7 8 9]


# Get the majoritiy of each class ( %)



In [0]:
def percentagecalculator(labels_cluster, c):
    cluster = stats.mode(labels_cluster)
    per = float(cluster[1]) / float(labels_cluster.shape[0]) * 100
    print("Cluster " + str(c) + " contains " + str(round(per, 2)) + " % number " + str(cluster[0]))

In [26]:
percentagecalculator(labels_cluster_0, 0)
percentagecalculator(labels_cluster_1, 1)
percentagecalculator(labels_cluster_2, 2)
percentagecalculator(labels_cluster_3, 3)
percentagecalculator(labels_cluster_4, 4)
percentagecalculator(labels_cluster_5, 5)
percentagecalculator(labels_cluster_6, 6)
percentagecalculator(labels_cluster_7, 7)
percentagecalculator(labels_cluster_8, 8)
percentagecalculator(labels_cluster_9, 9)

Cluster 0 contains 90.99 % number [6]
Cluster 1 contains 41.28 % number [7]
Cluster 2 contains 67.82 % number [1]
Cluster 3 contains 63.28 % number [8]
Cluster 4 contains 93.71 % number [0]
Cluster 5 contains 49.73 % number [3]
Cluster 6 contains 35.05 % number [9]
Cluster 7 contains 34.43 % number [0]
Cluster 8 contains 91.44 % number [2]
Cluster 9 contains 45.35 % number [4]


In this overview we see the most frequently occurring numbers in each cluster for each iteration (10,100,1000)

**Result for iterations = 10**

Cluster 1 contains 88.77 % number [6]


Cluster 2 contains 40.78 % number [7]


Cluster 3 contains 68.25 % number [1]


Cluster 4 contains 45.34 % number [8]


Cluster 5 contains 94.62 % number [0]


Cluster 6 contains 51.56 % number [3]


Cluster 7 contains 34.77 % number [9]


Cluster 8 contains 29.06 % number [5]


Cluster 9 contains 92.17 % number [2]


Cluster 10 contains 40.51 % number [4]

There are no cluster for number 4

**Result for iteration = 100**

Cluster 1 contains 90.99 % number [6]


Cluster 2 contains 41.28 % number [7]


Cluster 3 contains 67.82 % number [1]


Cluster 4 contains 63.28 % number [8]


Cluster 5 contains 93.71 % number [0]


Cluster 6 contains 49.73 % number [3]


Cluster 7 contains 35.05 % number [9]


Cluster 8 contains 34.43 % number [0]


Cluster 9 contains 91.44 % number [2]


Cluster 10 contains 45.35 % number [4]



There is no cluster for number 5, but we can see number 0 twice

**Result for iteration = 1000**

Cluster 1 contains 90.99 % number [6]


Cluster 2 contains 41.28 % number [7]


Cluster 3 contains 67.82 % number [1]


Cluster 4 contains 63.28 % number [8]


Cluster 5 contains 93.71 % number [0]


Cluster 6 contains 49.73 % number [3]


Cluster 7 contains 35.05 % number [9]


Cluster 8 contains 34.43 % number [0]


Cluster 9 contains 91.44 % number [2]


Cluster 10 contains 45.35 % number [4]

There is no cluster for number 5, but we can see number 0 twice